docs/architecture/adr-002-configv2.md

```diff
@@ -0,0 +1,109 @@
+The examples provided in this ADR are currently not exhaustive. So do NOT pay too much attention to each field, they're meant to be more examples of the _types_ of configuration for each object.
+
+## Context
+
+The configration of k0s has organically growed from very small to something quite complex today. While that is natural we're starting to see growing number of issues stemmin from the complexity.
+
+- Lack of separation between per node configuration and cluster configuration
+- Lack of feedback in dynamic config reconciliation status
+- Lack of feedback in helm and stack applier status (should these be in the clusterconfig to begin with?)
+- No separation between used defined configuration and the configuration that is used internally
+- Lack of versioning
+- Dynamic config has several problems when it comes to reconciliation of certain fields
+- Some fields have side effects that aren't very predictable.
+- The k0s' configuration for calico doesn't really match the calico configuration
+- Config doesn't allow to disable components, it requires modifying the cmdline, specially with dynamic config, it would make sense to allow this from the configuration.
+- In dynamicConfig it's not very straightforward what can and what cannot be modified
+
+All of this results in confusions on both the uers side and also on maintainers.
+
+## Config v2 goals
+
+The maintainers have agreed on set of high level goals for config v2:
+
+### Config as Kubernetes objects
+
+Configs shall be formatted as Kubernetes object, even if we do not really store them on API. K8s object are the natural language for k0s users and it keeps the option open to actually store everything in the API.
+
+### Per node and cluster wide config separation
+
+Separate the per node configuration and cluster-wide configs into their own CRDs. This makes it clear for both the users and maintainers where to look for which config data. It also makes clear separation on which _things_ can be changed at runtime via dynamic config.
+
+As an example, we would have `ControllerConfig` and `ClusterConfig` for the controlplanes:
+
+```yaml
+# ControllerConfig contains only the node specific bits for a controller node
+# So basically only the bits that we need to boot up etcd/kine and the API server
+apiVersion: k0sproject.io/...
+name: k0s
+kind: ControllerConfig
+spec:
+  etcd:
+    privateAddress: 1.2.3.4
+  api:
+    listenAddress: 1.2.3.4:6443,[::0]:6443
+    address: foobar.com:6443
+    sans:
+      - foobar.com
+  enableWorker: true
+  disableComponents: ["foo", "bar"]
+status:
+  lastReconciledConfig: "<ClusterConfig version: "Full object">"
+  history: # some kind of history when was which config applied etc.
+---
+# ClusterConfig contains all the bits that are always cluster-wide
+apiVersion: k0sproject.io/...
+name: k0s
+kind: ClusterConfig
+spec:
+  network:
+    provider: calico
+    calico:
+      foo: bar
+```
+
+Similarly we need to have config object for the workers. An example:
```

Comment:
Is it proposed here that we will have **one** `ControllerConfig` for all Controllers and **one** `WorkerConfig` for all Workers ? Wouldn't be useful to have some kind of "label" matching so one could apply certain configuration just for **some** of their workers for instance ?

One use case I can imagine: some of my worker nodes have GPU and on those I want to setup `containerd` differently because reasons.

Reply:
Hi Ricardo,

No, the idea is that we have one `ControllerConfig` for each controller and one `WorkerConfig` for each worker and one ClusterConfig for the whole cluster.

We also discussed the need of having one way of having `WorkerConfigs` templates, it can be a field in the ClusterConfig spec or it can be a new object called `WorkerGroup` or something similar.

My understanding is that in the ADR we shouldn't be very explicit on whether it should be a new object or not, but now that you mention it we should probably document having worker templates/groups as a goal.